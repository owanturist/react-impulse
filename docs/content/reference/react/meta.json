{
  "pages": [
    "use-scope",
    "use-scoped",
    "use-scoped-memo",
    "use-scoped-callback",
    "use-scoped-effect",
    "use-scoped-layout-effect"
  ]
}
