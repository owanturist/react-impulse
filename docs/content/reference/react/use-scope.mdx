---
title: useScope
description: API documentation for the useScope React hook
---

## Overview

The `useScope` hook returns a reactive scope tied to the component lifecycle.

## Function Signature

```typescript
function useScope(): Scope
```

## Returns

A `Scope` instance that triggers component re-renders when read impulses change.

## Description

`useScope` provides a scope that enables reactive impulse reads within React components. When impulses are read within this scope, the component automatically re-renders when those impulses change.

## Examples

```typescript
function Counter() {
  const scope = useScope()
  const impulse = Impulse(0)

  return (
    <div>
      Count: {impulse.getValue(scope)}
    </div>
  )
}
```