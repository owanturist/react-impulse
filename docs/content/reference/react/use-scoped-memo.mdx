---
title: useScopedMemo
description: API documentation for the useScopedMemo React hook
---

## Overview

The `useScopedMemo` hook memoizes computations that depend on impulses, similar to `React.useMemo` but with reactive scope support.

## Function Signature

```typescript
function useScopedMemo<T>(
  factory: (scope: Scope) => T,
  dependencies: DependencyList,
): T
```

## Parameters

- `factory`: A function that receives a `Scope` and computes the memoized value
- `dependencies`: An array of dependencies that control when the memoization is invalidated

## Returns

The memoized value of type `T`.

## Description

`useScopedMemo` combines impulse reactivity with React's memoization. The factory function is re-executed when either the dependencies change or when any impulses read within the scope change.

## Examples

```typescript
const memoizedValue = useScopedMemo(
  (scope) => expensiveComputation(impulse.getValue(scope)),
  [someDependency],
)
```
