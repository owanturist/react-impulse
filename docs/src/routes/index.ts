import { createFileRoute } from "@tanstack/react-router"

import { Home } from "~/pages/home"

export const Route = createFileRoute("/")({
  component: Home,
})
