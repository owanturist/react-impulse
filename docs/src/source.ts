export { create, docs as source } from "../source.generated"
