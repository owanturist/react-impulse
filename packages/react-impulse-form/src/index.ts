export * from "./impulse-form"

export * from "./impulse-form-unit"
export * from "./impulse-form-shape"
export type * from "./impulse-form-meta"
export * from "./impulse-form-list"
export * from "./impulse-form-switch"
export * from "./impulse-form-optional"

export type * from "./result"
export type * from "./zod-like-schema"
export type { ValidateStrategy } from "./validate-strategy"
