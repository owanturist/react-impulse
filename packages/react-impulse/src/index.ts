export type { DependencyList } from "./dependencies"
export type { Compare, Destructor } from "./utils"
export {
  type ImpulseOptions,
  type TransmittingImpulseOptions,
  type ReadonlyImpulse,
  type ImpulseGetter,
  Impulse,
} from "./Impulse"
export type { Scope } from "./Scope"
export { untrack } from "./untrack"
export { batch } from "./batch"
export { batch as tap } from "./batch"
export { subscribe } from "./subscribe"
export { useScope } from "./useScope"
export { type UseScopedOptions, useScoped } from "./useScoped"
export { useScopedMemo } from "./useScopedMemo"
export { useScopedCallback } from "./useScopedCallback"
export { useScopedEffect } from "./useScopedEffect"
export { useScopedLayoutEffect } from "./useScopedLayoutEffect"
