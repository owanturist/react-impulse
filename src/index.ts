export type { Compare } from "./utils"
export { Impulse } from "./Sweety"
export { batch } from "./batch"
export { watch } from "./watch"
export { useImpulse } from "./useSweety"
export { useImpulseValue } from "./useSweetyState"
export { useImpulseMemo } from "./useImpulseMemo"
export { useImpulseEffect, useImpulseLayoutEffect } from "./useImpulseEffects"
export { useWatchImpulse } from "./useWatchImpulse"
