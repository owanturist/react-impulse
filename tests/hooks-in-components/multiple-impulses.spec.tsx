import React from "react"
import { act, render, screen, fireEvent } from "@testing-library/react"

import { Impulse, useImpulseValue } from "../../src"

describe("multiple impulses", () => {
  const LoginForm: React.FC<{
    email: Impulse<string>
    password: Impulse<string>
    onRender: VoidFunction
  }> = ({ email: emailImpulse, password: passwordImpulse, onRender }) => {
    const email = useImpulseValue(emailImpulse)
    const password = useImpulseValue(passwordImpulse)

    return (
      <React.Profiler id="test" onRender={onRender}>
        <input
          type="email"
          data-testid="email"
          value={email}
          onChange={(event) => emailImpulse.setValue(event.target.value)}
        />
        <input
          type="password"
          data-testid="password"
          value={password}
          onChange={(event) => passwordImpulse.setValue(event.target.value)}
        />
        <button
          type="button"
          data-testid="reset"
          onClick={() => {
            emailImpulse.setValue("")
            passwordImpulse.setValue("")
          }}
        />
      </React.Profiler>
    )
  }

  it("performs multi impulse management", () => {
    const email = Impulse.of("")
    const password = Impulse.of("")
    const onRender = vi.fn()

    const { container } = render(
      <LoginForm email={email} password={password} onRender={onRender} />,
    )

    expect(onRender).toHaveBeenCalledOnce()
    expect(container).toMatchSnapshot()
    vi.clearAllMocks()

    // change email
    fireEvent.change(screen.getByTestId("email"), {
      target: { value: "john-doe@gmail.com" },
    })
    expect(onRender).toHaveBeenCalledOnce()
    expect(container).toMatchSnapshot()
    vi.clearAllMocks()

    // change password
    fireEvent.change(screen.getByTestId("password"), {
      target: { value: "qwerty" },
    })
    expect(onRender).toHaveBeenCalledOnce()
    expect(container).toMatchSnapshot()
    vi.clearAllMocks()

    // changes from the outside
    act(() => {
      email.setValue("admin@gmail.com")
      password.setValue("admin")
    })
    expect(onRender).toHaveBeenCalledOnce()
    expect(container).toMatchSnapshot()
    vi.clearAllMocks()

    // reset
    fireEvent.click(screen.getByTestId("reset"))
    expect(onRender).toHaveBeenCalledOnce()
    expect(container).toMatchSnapshot()
  })
})
