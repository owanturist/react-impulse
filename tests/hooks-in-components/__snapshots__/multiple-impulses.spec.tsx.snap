// Vitest Snapshot v1

exports[`multiple impulses > Performs multi impulse management 1`] = `
<div>
  <input
    data-testid="email"
    type="email"
    value=""
  />
  <input
    data-testid="password"
    type="password"
    value=""
  />
  <button
    data-testid="reset"
    type="button"
  />
</div>
`;

exports[`multiple impulses > Performs multi impulse management 2`] = `
<div>
  <input
    data-testid="email"
    type="email"
    value="john-doe@gmail.com"
  />
  <input
    data-testid="password"
    type="password"
    value=""
  />
  <button
    data-testid="reset"
    type="button"
  />
</div>
`;

exports[`multiple impulses > Performs multi impulse management 3`] = `
<div>
  <input
    data-testid="email"
    type="email"
    value="john-doe@gmail.com"
  />
  <input
    data-testid="password"
    type="password"
    value="qwerty"
  />
  <button
    data-testid="reset"
    type="button"
  />
</div>
`;

exports[`multiple impulses > Performs multi impulse management 4`] = `
<div>
  <input
    data-testid="email"
    type="email"
    value="admin@gmail.com"
  />
  <input
    data-testid="password"
    type="password"
    value="admin"
  />
  <button
    data-testid="reset"
    type="button"
  />
</div>
`;

exports[`multiple impulses > Performs multi impulse management 5`] = `
<div>
  <input
    data-testid="email"
    type="email"
    value=""
  />
  <input
    data-testid="password"
    type="password"
    value=""
  />
  <button
    data-testid="reset"
    type="button"
  />
</div>
`;
